src/start/LoadingScreen.js

```diff
@@ -15,15 +19,95 @@ const styles = createStyleSheet({
   },
 });
 
+type SelectorProps = {|
+  hasAuth: boolean,
+  accounts: Account[],
+  users: User[],
+  isHydrated: boolean,
+|};
+
 type Props = $ReadOnly<{|
   // Since we've put this screen in a stack-nav route config, and we
   // don't invoke it without type-checking anywhere else (in fact, we
   // don't invoke it anywhere else at all), we know it gets the
   // `navigation` prop for free, with the stack-nav shape.
   navigation: NavigationStackProp<NavigationStateRoute>,
+
+  dispatch: Dispatch,
+  ...SelectorProps,
 |}>;
 
-export default class LoadingScreen extends PureComponent<Props> {
+class LoadingScreen extends PureComponent<Props> {
+  static hasDoneInitialNavigation = false;
+
+  /**
+   * Do initial navigation unless it's already been done.
+   */
+  maybeDoInitialNavigation = () => {
+    if (LoadingScreen.hasDoneInitialNavigation || !this.props.isHydrated) {
+      return;
+    }
+    this.doInitialNavigation();
+  };
+
+  /**
+   * Data has been loaded, so open the app to the right screen.
+   *
+   * Not to be called before the REHYDRATE action, and not to be
+   * called more than once.
+   */
+  doInitialNavigation = () => {
```

Comment:
(We discussed this commit today by video chat. One thing we figured out is that the use of a ref on the navigator means that it belongs in an ancestor component of the navigator, not a descendant like `LoadingScreen`.)